### PR TITLE
fix populateClusterRoleBindingNames bug

### PR DESCRIPTION
**Pull request by sam-nguyen7**
Bug: Comparing two []string and output difference onto an array. Nested for loop did not achieve that

### CODE REVIEW

File: pkg/controller/grouppermission/grouppermission_controller.go [hunk 1]
```diff
@@ -203,18 +203,23 @@ func populateCrClusterRoleNames(groupPermission *managedv1alpha1.GroupPermission
...
 	// get ClusterRoleBindingList
 	onClusterItems := clusterRoleBindingList.Items
 
-	for _, i := range onClusterItems {
-		for _, a := range clusterRoleBindingName {
-			if a != i.Name {
-				crClusterRoleBindingList = append(crClusterRoleBindingList, a)
-			}
+	// map onClusterItems
+	mb := make(map[string]struct{}, len(onClusterItems))
```
Comment (lisa) [inline pkg/controller/grouppermission/grouppermission_controller.go]: Why is this a `struct{}`? What is the type of `clusterRoleBindingList.Items`?

File: pkg/controller/grouppermission/grouppermission_controller.go [hunk 2]
```diff
@@ -203,18 +203,23 @@ func populateCrClusterRoleNames(groupPermission *managedv1alpha1.GroupPermission
 }
 
 // populateClusterRoleBindingNames to see if ClusterRoleBinding exists in k8s ClusterRoleBindlingList
-// returns bool
-func populateClusterRoleBindingNames(clusterRoleBindingName []string, clusterRoleBindingList *v1.ClusterRoleBindingList) []string {
+// returns list of clusterRoleBinding
```
Comment (jharrington22) [inline pkg/controller/grouppermission/grouppermission_controller.go]: comment doesn't reflect what is actually being returned

File: pkg/controller/grouppermission/grouppermission_controller.go [hunk 3]
```diff
@@ -203,18 +203,23 @@ func populateCrClusterRoleNames(groupPermission *managedv1alpha1.GroupPermission
...
+func populateClusterRoleBindingNames(clusterRoleBindingNames []string, clusterRoleBindingList *v1.ClusterRoleBindingList) []string {
```
Comment (jharrington22) [inline pkg/controller/grouppermission/grouppermission_controller.go]: @sam-nguyen7 I would think something like the following would be easier to read what do you think?

```
var missingRoleBindings []string
var found bool

if len(clusterRoleBindingList.Items) == len(clusterRoleBindingNames) {
	return missingRoleBindings
}

for _, crbn := range clusterRoleBindingNames {
        for _, crbl := range clusterRoleBindingList.Items {
		if crbn == crbl {
			found = true
		}
	}
	if found {
		missingRoleBindings = append(missingRoleBindings, crbn)
	}
	found = false
}

return missingRoleBindings
```

Comment (jharrington22) [reply pkg/controller/grouppermission/grouppermission_controller.go]: @lisa would you think so? There might be a more elegant solution

Comment (fahlmant) [reply pkg/controller/grouppermission/grouppermission_controller.go]: I would do this, except for the length check at the start. If an item is removed and another created in the list of desired objects, then the length would be equal, but the contents would be different.

File: pkg/controller/grouppermission/grouppermission_controller.go [hunk 4]
```diff
@@ -203,21 +203,25 @@ func populateCrClusterRoleNames(groupPermission *managedv1alpha1.GroupPermission
 }
 
 // populateClusterRoleBindingNames to see if ClusterRoleBinding exists in k8s ClusterRoleBindlingList
-// returns bool
-func populateClusterRoleBindingNames(clusterRoleBindingName []string, clusterRoleBindingList *v1.ClusterRoleBindingList) []string {
+// returns list of clusterRoleBindingNames that exists in CR but not in clusterRoleBindingList
```
Comment (jharrington22) [inline pkg/controller/grouppermission/grouppermission_controller.go]: returns a slice :)

File: pkg/controller/grouppermission/grouppermission_controller.go [hunk 5]
```diff
@@ -203,21 +203,25 @@ func populateCrClusterRoleNames(groupPermission *managedv1alpha1.GroupPermission
...
 
 	// get ClusterRoleBindingList
 	onClusterItems := clusterRoleBindingList.Items
 
-	for _, i := range onClusterItems {
-		for _, a := range clusterRoleBindingName {
-			if a != i.Name {
-				crClusterRoleBindingList = append(crClusterRoleBindingList, a)
+	for _, crbName := range clusterRoleBindingNames {
+		for _, crBinding := range onClusterItems {
```
Comment (jharrington22) [inline pkg/controller/grouppermission/grouppermission_controller.go]: I'd be inclined here to just call `clusterRoleBindingList.Items`

File: pkg/controller/grouppermission/grouppermission_controller.go (hunk 4 above)
Comment (sam-nguyen7) [reply pkg/controller/grouppermission/grouppermission_controller.go]: ah yes.

File: pkg/controller/grouppermission/grouppermission_controller.go [hunk 6]
```diff
@@ -203,21 +203,22 @@ func populateCrClusterRoleNames(groupPermission *managedv1alpha1.GroupPermission
 }
 
 // populateClusterRoleBindingNames to see if ClusterRoleBinding exists in k8s ClusterRoleBindlingList
-// returns bool
-func populateClusterRoleBindingNames(clusterRoleBindingName []string, clusterRoleBindingList *v1.ClusterRoleBindingList) []string {
+// returns a slice of clusterRoleBindingNames that exists in CR but not in clusterRoleBindingList
+func populateClusterRoleBindingNames(clusterRoleBindingNames []string, clusterRoleBindingList *v1.ClusterRoleBindingList) []string {
 	var crClusterRoleBindingList []string
+	var found bool
```
Comment (lisa) [inline pkg/controller/grouppermission/grouppermission_controller.go]:
```suggestion
	found := false
```

It may be safer to initialize the value?

**Comment by lisa** [comment]
/lgtm

Will leave @jharrington22 to clear his requested changes.

File: pkg/controller/grouppermission/grouppermission_controller.go (hunk 6 above)
Comment (sam-nguyen7) [reply pkg/controller/grouppermission/grouppermission_controller.go]: @jharrington22 default for bool is false I believe.

**Pull request merged**